text/0031-integ-testing.md

```diff
@@ -0,0 +1,542 @@
+executed as part of an AWS CloudFormation stack deployment, they take time to execute.
+It will not be possible to, deploy the app once and run assertions many times against it.
+
+These insufficiencies could lead users to make poor choices on how they model their tests.
+For example, bunch unrelated assertions together in the same test in order to reduce test
+execution time.
+
+### Alternative - run assertions locally
+
+Alternatively, the test runner executes these assertions locally. Either the developer machine
+or the CI server will interact directly with the deployed CDK app to execute the assertions.
+
+The proposed approach is better than this alternative for the following reasons.
+
+The proposed solution allows for users, typically enterprise users, to test "private" resources,
+such as a lambda function in a private vpc. With this alternative, the resource will need to be
+exposed to the developer machine and/or CI server.
+
+This type of testing being proposed here involves deploying and cleaning up AWS resources via
+CloudFormation stacks for each test. This is time intensive. A combined snapshot of the CDK app
+and assertions, taken once, provides a high level of confidence that the test does not need to be
+re-run. A fresh snapshot is taken every time the CDK app or any of the assertions in the test
+changes.
+Since the proposed solution models the assertions as AWS CDK constructs, the synthesized [cloud
+assembly] functions as a full and effective snapshot.
+This alternative, on the other hand, will require inventing an effective snapshotting solution.
+
+The proposed solution also provides a consistent execution environment for tests - the [AWS Lambda
+base images] and the ability to [modify the environment][Modifying runtime environment] via
+extensions and layers.
+As an example, if an assertion depends on the `sed` utility. OSX ships by default with BSD sed,
+while Linux ships with GNU sed, which have subtly different behaviours and option switches.
+In the proposed solution, the assertion will either use the `sed` utility in the Lambda base
+image or bring its own via Lambda Layers.
+If this is not accounted for, this alternative will cause the assertion to behave inconsistently.
+
+[AWS CloudFormation custom resources]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-custom-resources.html
+[cloud assembly]: https://docs.aws.amazon.com/cdk/api/latest/docs/cloud-assembly-schema-readme.html#cloud-assembly
+[AWS Lambda base images]: https://docs.aws.amazon.com/lambda/latest/dg/runtimes-images.html
+[Modifying runtime environment]: https://docs.aws.amazon.com/lambda/latest/dg/runtimes-modify.html
+
+## Appendix B - Test Execution
+
+### Design
+
+The AWS CDK CLI (aka toolkit) is responsible for deploying and destroying AWS CDK apps.
+We will extract these two parts from the CLI into a separate jsii module.
+
+The `run()` method on the class `apptest.Test` will depend on this module to deploy and
```

Comment:
Feels like this is a bit of a rabbit hole not directly needed to support integration tests. Why can't integration tests be simple CDK apps (like `cdk-integ` works) and then we can just use the CDK CLI (or a tool that wraps the CDK CLI) to deploy them?